compiler/rustc_resolve/src/rustdoc.rs

```diff
@@ -528,6 +536,38 @@ pub fn source_span_for_markdown_range(
     let is_all_sugared_doc = fragments.iter().all(|frag| frag.kind == DocFragmentKind::SugaredDoc);
 
     if !is_all_sugared_doc {
+        // this case ignores the markdown outside of the range so that it can
+        // work in cases where the markdown is made from several different
+        // doc fragments, but the target range does not span across multiple
+        // fragments.
+        let mut match_data = None;
+        let pat = &markdown[md_range.clone()];
+        // this heirustic doesn't make sense with a zero-sized range.
+        if pat.len() == 0 {
+            return None;
+        }
+        for (i, fragment) in fragments.iter().enumerate() {
+            if let Ok(snippet) = span_to_snippet(fragment.span)
+                && let Some(match_start) = snippet.find(pat)
+            {
+                // if there is either a match in a previous fragment,
+                // or multiple matches in this fragment,
+                // there is ambiguity.
+                if match_data.is_none() && !snippet[match_start + 1..].contains(pat) {
+                    match_data = Some((i, match_start));
+                } else {
+                    // heirustic produced ambiguity, return nothing.
+                    return None;
+                }
+            }
+        }
+        if let Some((i, match_start)) = match_data {
+            use rustc_span::BytePos;
+            let mut sp = fragments[i].span;
+            sp = sp.with_lo(sp.lo() + BytePos(match_start as u32));
+            sp = sp.with_hi(sp.lo() + BytePos((md_range.end - md_range.start) as u32));
+            return Some(sp);
```

Comment:
```suggestion
            return Some(fragments[i].span.with_lo(sp.lo() + BytePos(match_start as u32)).with_hi(sp.lo() + BytePos((md_range.end - md_range.start) as u32)));
```

Reply:
this actually meaningfully changes the semantics drastically.

the reason it's done in 2 starts is we need to mutate `sp.lo` and then use the *new* `sp.lo` in the calculation for `sp.hi`

Reply:
That looks.. ok? It does lose all the pleasing alignment from the imperative solution, its quite nested so I'm not sure its any easier to read, and also, I've already iterated on this section like 5 times, and I'm not particularly keen to rewrite it again.

Reply:
yeahhh alright i'll rewrite it again, what's one more on the pile.